// better-auth/dist/index.mjs
import { getBaseURL, getHost, getHostFromRequest, getOrigin, getProtocol, getProtocolFromRequest, isDynamicBaseURLConfig, matchesHostPattern, resolveBaseURL, resolveDynamicBaseURL } from "./utils/url.mjs";
import { generateGenericState, parseGenericState } from "./state.mjs";
import { generateState, parseState } from "./oauth2/state.mjs";
import { HIDE_METADATA } from "./utils/hide-metadata.mjs";
import "./utils/index.mjs";
import { APIError } from "./api/index.mjs";
import { betterAuth } from "./auth/full.mjs";
import { getCurrentAdapter } from "@better-auth/core/context";
import { createTelemetry, getTelemetryAuthConfig } from "@better-auth/telemetry";

export * from "@better-auth/core"

export * from "@better-auth/core/db"

export * from "@better-auth/core/env"

export * from "@better-auth/core/error"

export * from "@better-auth/core/oauth2"

export * from "@better-auth/core/utils/error-codes"

export * from "@better-auth/core/utils/id"

export * from "@better-auth/core/utils/json"

export { APIError, HIDE_METADATA, betterAuth, createTelemetry, generateGenericState, generateState, getBaseURL, getCurrentAdapter, getHost, getHostFromRequest, getOrigin, getProtocol, getProtocolFromRequest, getTelemetryAuthConfig, isDynamicBaseURLConfig, matchesHostPattern, parseGenericState, parseState, resolveBaseURL, resolveDynamicBaseURL };
