// better-auth/dist/plugins/index.mjs
import { HIDE_METADATA } from "../utils/hide-metadata.mjs";
import { createAccessControl, role } from "./access/access.mjs";
import "./access/index.mjs";
import { MULTI_SESSION_ERROR_CODES } from "./multi-session/error-codes.mjs";
import { TWO_FACTOR_ERROR_CODES } from "./two-factor/error-code.mjs";
import { twoFactorClient } from "./two-factor/client.mjs";
import { USERNAME_ERROR_CODES } from "./username/error-codes.mjs";
import { admin } from "./admin/admin.mjs";
import "./admin/index.mjs";
import { anonymous } from "./anonymous/index.mjs";
import { bearer } from "./bearer/index.mjs";
import { captcha } from "./captcha/index.mjs";
import { customSession } from "./custom-session/index.mjs";
import { deviceAuthorization, deviceAuthorizationOptionsSchema } from "./device-authorization/index.mjs";
import { emailOTP } from "./email-otp/index.mjs";
import { auth0 } from "./generic-oauth/providers/auth0.mjs";
import { gumroad } from "./generic-oauth/providers/gumroad.mjs";
import { hubspot } from "./generic-oauth/providers/hubspot.mjs";
import { keycloak } from "./generic-oauth/providers/keycloak.mjs";
import { line } from "./generic-oauth/providers/line.mjs";
import { microsoftEntraId } from "./generic-oauth/providers/microsoft-entra-id.mjs";
import { okta } from "./generic-oauth/providers/okta.mjs";
import { patreon } from "./generic-oauth/providers/patreon.mjs";
import { slack } from "./generic-oauth/providers/slack.mjs";
import { genericOAuth } from "./generic-oauth/index.mjs";
import { haveIBeenPwned } from "./haveibeenpwned/index.mjs";
import { createJwk, generateExportedKeyPair, toExpJWT } from "./jwt/utils.mjs";
import { getJwtToken, signJWT } from "./jwt/sign.mjs";
import { verifyJWT } from "./jwt/verify.mjs";
import { jwt } from "./jwt/index.mjs";
import { lastLoginMethod } from "./last-login-method/index.mjs";
import { magicLink } from "./magic-link/index.mjs";
import { getClient, getMetadata, oidcProvider } from "./oidc-provider/index.mjs";
import { getMCPProtectedResourceMetadata, getMCPProviderMetadata, mcp, oAuthDiscoveryMetadata, oAuthProtectedResourceMetadata, withMcpAuth } from "./mcp/index.mjs";
import { multiSession } from "./multi-session/index.mjs";
import { oAuthProxy } from "./oauth-proxy/index.mjs";
import { oneTap } from "./one-tap/index.mjs";
import { oneTimeToken } from "./one-time-token/index.mjs";
import { openAPI } from "./open-api/index.mjs";
import { getOrgAdapter } from "./organization/adapter.mjs";
import { hasPermission } from "./organization/has-permission.mjs";
import { organization, parseRoles } from "./organization/organization.mjs";
import "./organization/index.mjs";
import { phoneNumber } from "./phone-number/index.mjs";
import { siwe } from "./siwe/index.mjs";
import { testUtils } from "./test-utils/index.mjs";
import { twoFactor } from "./two-factor/index.mjs";
import { username } from "./username/index.mjs";

export { MULTI_SESSION_ERROR_CODES as ERROR_CODES, HIDE_METADATA, TWO_FACTOR_ERROR_CODES, USERNAME_ERROR_CODES, admin, anonymous, auth0, bearer, captcha, createAccessControl, createJwk, customSession, deviceAuthorization, deviceAuthorizationOptionsSchema, emailOTP, generateExportedKeyPair, genericOAuth, getClient, getJwtToken, getMCPProtectedResourceMetadata, getMCPProviderMetadata, getMetadata, getOrgAdapter, gumroad, hasPermission, haveIBeenPwned, hubspot, jwt, keycloak, lastLoginMethod, line, magicLink, mcp, microsoftEntraId, multiSession, oAuthDiscoveryMetadata, oAuthProtectedResourceMetadata, oAuthProxy, oidcProvider, okta, oneTap, oneTimeToken, openAPI, organization, parseRoles, patreon, phoneNumber, role, signJWT, siwe, slack, testUtils, toExpJWT, twoFactor, twoFactorClient, username, verifyJWT, withMcpAuth };
