// better-auth/dist/client/plugins/index.mjs
import { inferAdditionalFields } from "../../plugins/additional-fields/client.mjs";
import { ADMIN_ERROR_CODES } from "../../plugins/admin/error-codes.mjs";
import { adminClient } from "../../plugins/admin/client.mjs";
import { ANONYMOUS_ERROR_CODES } from "../../plugins/anonymous/error-codes.mjs";
import { anonymousClient } from "../../plugins/anonymous/client.mjs";
import { customSessionClient } from "../../plugins/custom-session/client.mjs";
import { deviceAuthorizationClient } from "../../plugins/device-authorization/client.mjs";
import { EMAIL_OTP_ERROR_CODES } from "../../plugins/email-otp/error-codes.mjs";
import { emailOTPClient } from "../../plugins/email-otp/client.mjs";
import { GENERIC_OAUTH_ERROR_CODES } from "../../plugins/generic-oauth/error-codes.mjs";
import { genericOAuthClient } from "../../plugins/generic-oauth/client.mjs";
import { jwtClient } from "../../plugins/jwt/client.mjs";
import { lastLoginMethodClient } from "../../plugins/last-login-method/client.mjs";
import { magicLinkClient } from "../../plugins/magic-link/client.mjs";
import { MULTI_SESSION_ERROR_CODES } from "../../plugins/multi-session/error-codes.mjs";
import { multiSessionClient } from "../../plugins/multi-session/client.mjs";
import { oidcClient } from "../../plugins/oidc-provider/client.mjs";
import { oneTapClient } from "../../plugins/one-tap/client.mjs";
import { oneTimeTokenClient } from "../../plugins/one-time-token/client.mjs";
import { ORGANIZATION_ERROR_CODES } from "../../plugins/organization/error-codes.mjs";
import { clientSideHasPermission, inferOrgAdditionalFields, organizationClient } from "../../plugins/organization/client.mjs";
import { PHONE_NUMBER_ERROR_CODES } from "../../plugins/phone-number/error-codes.mjs";
import { phoneNumberClient } from "../../plugins/phone-number/client.mjs";
import { siweClient } from "../../plugins/siwe/client.mjs";
import { TWO_FACTOR_ERROR_CODES } from "../../plugins/two-factor/error-code.mjs";
import { twoFactorClient } from "../../plugins/two-factor/client.mjs";
import { USERNAME_ERROR_CODES } from "../../plugins/username/error-codes.mjs";
import { usernameClient } from "../../plugins/username/client.mjs";
import { InferServerPlugin } from "./infer-plugin.mjs";

export { ADMIN_ERROR_CODES, ANONYMOUS_ERROR_CODES, EMAIL_OTP_ERROR_CODES, GENERIC_OAUTH_ERROR_CODES, InferServerPlugin, MULTI_SESSION_ERROR_CODES, ORGANIZATION_ERROR_CODES, PHONE_NUMBER_ERROR_CODES, TWO_FACTOR_ERROR_CODES, USERNAME_ERROR_CODES, adminClient, anonymousClient, clientSideHasPermission, customSessionClient, deviceAuthorizationClient, emailOTPClient, genericOAuthClient, inferAdditionalFields, inferOrgAdditionalFields, jwtClient, lastLoginMethodClient, magicLinkClient, multiSessionClient, oidcClient, oneTapClient, oneTimeTokenClient, organizationClient, phoneNumberClient, siweClient, twoFactorClient, usernameClient };
